migrations/kdvh/import_functions.go

```diff
@@ -0,0 +1,463 @@
+//  9 - SECOND DIGIT OF HEXADECIMAL VALUE OF THE OBSERVATION CONFIDENCE LEVEL
+// 13 - FIRST HQC OPERATOR DIGIT
+// 14 - SECOND HQC OPERATOR DIGIT
+// 15 - HEXADECIMAL DIGIT WITH NUMBER OF TESTS THAT DID NOT PASS (RETURNED A RESULT?)
+
+const (
+	// Remaing 11 digits of `useinfo` that follow the 5 digits contained in `obs.Flags`.
+	// TODO: From the docs it looks like the '9' should be changed by kvalobs when
+	// the observation is inserted into the database but that's not the case?
+	DELAY_DEFAULT = "00900000000"
+
+	INVALID_FLAGS                = "99999" + DELAY_DEFAULT // Only returned when the flags are invalid
+	COMPLETED_HQC                = "40000" + DELAY_DEFAULT // Specific to T_VDATA
+	DIURNAL_INTERPOLATED_USEINFO = "48925" + DELAY_DEFAULT // Specific to T_DIURNAL_INTERPOLATED
+)
+
+func (obs *Obs) flagsAreValid() bool {
+	if len(obs.Flags) != 5 {
+		return false
+	}
+	_, err := strconv.ParseInt(obs.Flags, 10, 32)
+	return err == nil
+}
+
+func (obs *Obs) Useinfo() string {
+	if !obs.flagsAreValid() {
+		return INVALID_FLAGS
+	}
+	return obs.Flags + DELAY_DEFAULT
+}
+
+// The following functions try to recover the original pair of `controlinfo`
+// and `useinfo` generated by Kvalobs for the observation, based on `Obs.Flags` and `Obs.Data`
+// Different KDVH tables need different ways to perform this conversion.
+
+func makeDataPage(obs Obs) (LardObs, error) {
+	var valPtr *float32
+
+	controlinfo := VALUE_PASSED_QC
+	if obs.Data == "" {
+		controlinfo = VALUE_MISSING
+	}
+
+	// NOTE: this is the only function that can return `LardObs`
+	// with non-null text data
+	if !obs.param.IsScalar {
+		return LardObs{
+			Obstime:     obs.Obstime,
+			Text:        &obs.Data,
+			Useinfo:     obs.Useinfo(),
+			Controlinfo: controlinfo,
+		}, nil
+	}
+
+	val, err := strconv.ParseFloat(obs.Data, 32)
+	if err == nil {
+		f32 := float32(val)
+		valPtr = &f32
+	}
+
+	return LardObs{
+		Obstime:     obs.Obstime,
+		Data:        valPtr,
+		Useinfo:     obs.Useinfo(),
+		Controlinfo: controlinfo,
+	}, nil
+}
+
+// modify obstimes to always use totime
+func makeDataPageProduct(obs Obs) (LardObs, error) {
+	obsLard, err := makeDataPage(obs)
+	if !obs.offset.IsZero() {
+		if temp, ok := obs.offset.AddTo(obsLard.Obstime); ok {
+			obsLard.Obstime = temp
+		}
+	}
+	return obsLard, err
+}
+
+func makeDataPageEdata(obs Obs) (LardObs, error) {
+	var controlinfo string
+	var valPtr *float32
+
+	if val, err := strconv.ParseFloat(obs.Data, 32); err != nil {
+		switch obs.Flags {
+		case "70381", "70389", "90989":
+			controlinfo = VALUE_REMOVED_BY_QC
+		default:
+			// Includes "70000", "70101", "99999"
+			controlinfo = VALUE_MISSING
+		}
+	} else {
+		controlinfo = VALUE_PASSED_QC
+		f32 := float32(val)
+		valPtr = &f32
+	}
+
+	return LardObs{
+		Obstime:     obs.Obstime,
+		Data:        valPtr,
+		Useinfo:     obs.Useinfo(),
+		Controlinfo: controlinfo,
+	}, nil
+}
+
+func makeDataPagePdata(obs Obs) (LardObs, error) {
```

Comment:
No, the distinction is only kept in the flags, since the `obsvalue` column in LARD can be NULL. To be honest I'm not even sure if I should be inserting NULLs, because when we query we can simply generate NULLs on the fly if data is missing.

Reply:
I see, so there are NULLs and the flags can be used to differentiate what the NULL means? I recommend asking Pål Sannes about this in case nobody has done so yet, to make sure that this information can be deduced (from kvalobs flags?) without those odd code values

Reply:
Yes, if I understand correctly you can distinguish the two cases from these two flags (maybe the comment description can be improved). I can ask him though
https://github.com/metno/lard/blob/2b5ebf9f1c34acbf9377a78b497f28a22c63f54e/migrations/kdvh/import_functions.go#L147-L148

Reply:
> No, the distinction is only kept in the flags, since the `obsvalue` column in LARD can be NULL. To be honest I'm not even sure if I should be inserting NULLs, because when we query we can simply generate NULLs on the fly if data is missing.

@intarga, do you have any opinion about inserting NULLs?

Reply:
I don't. I can see the argument for both approaches though. The difference here is that this data did exist and was deleted by QC, a state that can never happen in our new system, so that's the case for inserting a NULL.